### PR TITLE
add plaid-bar

### PR DESCRIPTION
**Pull request by kturney**
I'm looking to use maximum plaid in a new chart heavy app at work. The first chart I've added is bar. I'd love any feedback you have. I aligned it pretty closely to how plaid-axis currently works.

Before we merge it in, I'd like to add orientation so we can do horizontal bars charts as well.
Just want to start the conversation early :-)

### CODE REVIEW

**Comment by kturney** [comment]
I'm also going to need a stacked bar chart, so maybe that goes in here too.

File: addon/components/plaid-bar/component.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  drawBars() {
+    let { values, xScale, yScale, height, fill, fillOpacity } =
+      getProperties(this, 'values', 'xScale', 'yScale', 'height', 'fill', 'fillOpacity');
+
+    this.groupElement.selectAll('.bar').data(values).enter().append('rect')
+      .attr('class', 'bar')
+      .attr('x', (d) => xScale(d[0]))
+      .attr('width', () => xScale.bandwidth())
+      .attr('y', (d) => yScale(d[1]))
+      .attr('height', (d) => height - yScale(d[1]))
```
Comment (ivanvanderbyl) [inline addon/components/plaid-bar/component.js]: You could actually get away without it needing to know the height by using `yScale.range()[1]`, which has the added benefit that you can reverse the bars by reversing the range (same as the line chart)

File: tests/integration/components/plaid-bar/component-test.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+test('it renders', function(assert) {
+  // Set any properties with this.set('myProperty', 'value');
+  // Handle any actions with this.on('myAction', function(val) { ... });
+
+  this.render(hbs`{{plaid-bar}}`);
+
+  assert.equal(this.$().text().trim(), '');
+
+  // Template block usage:
+  this.render(hbs`
```
Comment (ivanvanderbyl) [inline tests/integration/components/plaid-bar/component-test.js:142]: We should add some proper tests, assert that each bar gets drawn, and has correct position and height.

File: addon/components/plaid-bar/component.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+  fill: 'black',
+
+  fillOpacity: 1.0,
+
+  height: null,
+
+  didInsertElement() {
+    this._super(...arguments);
+    this.groupElement = select(this.element);
```
Comment (ivanvanderbyl) [inline addon/components/plaid-bar/component.js]: The `GroupElement` mixin should already handle this.

File: addon/components/plaid-bar/component.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    scheduleOnce('afterRender', this, this.drawBars);
```
Comment (ivanvanderbyl) [inline addon/components/plaid-bar/component.js]: I don't think this is needed because `didReceiveAttrs` will fire anyway.

**Comment by ivanvanderbyl** [comment]
@kturney thanks for this, it's a great start. I think with a few tweaks we can ship it as the basic bar chart. With stacked bars, I think much like the stacked area chart we can use the stack layout, and somehow compose it, perhaps as a helper or a group:

``` hbs
{{stacked-layout values as |stackedValues| }}
  {{plot.bars stackedValues}}
{{/stacked-layout}}
```

Or

``` hbs
{{plot.bars (stacked values)}}
```

I haven't looked at which is actually practical yet.

File: addon/components/plaid-bar/component.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  },
+
+  drawBars() {
+    let { values, drawnValues, xScale, yScale, fill, fillOpacity, orientation } =
+      getProperties(this, 'values', 'drawnValues', 'xScale', 'yScale', 'fill', 'fillOpacity', 'orientation');
+
+    let x, width, y, height;
+
+    if (orientation === 'vertical') {
+      let maxHeight = Math.max(...yScale.range());
```
Comment (ivanvanderbyl) [inline addon/components/plaid-bar/component.js]: We should probably standardize on using `d3-array` min/max here, not that there's much efficiency gains on a 2 element array :D

**Pull request merged**